File: src/compiler/prepare_grammar/intern_symbols.cc
#include "compiler/prepare_grammar/intern_symbols.h"
#include <memory>
#include <vector>
#include <cassert>
#include <set>
#include "tree_sitter/compiler.h"
#include "compiler/grammar.h"
#include "compiler/rule.h"

namespace tree_sitter {
namespace prepare_grammar {

using std::string;
using std::vector;
using std::set;
using std::pair;
using rules::Symbol;
using rules::Rule;

class SymbolInterner {
 public:
  Rule apply(const Rule &rule) {
    return rule.match(
      [&](const rules::Blank &blank) -> Rule {
        return blank;
      },

      [&](const rules::NamedSymbol &symbol) {
        return intern_symbol(symbol);
      },

      [&](const rules::String &string) {
        return string;
      },

      [&](const rules::Pattern &pattern) {
        return pattern;
      },

      [&](const rules::Choice &choice) {
        vector<rules::Rule> elements;
        for (const auto &element : choice.elements) {
          elements.push_back(apply(element));
        }
        return rules::Choice{elements};
      },

      [&](const rules::Seq &sequence) {
        return rules::Seq{apply(*sequence.left), apply(*sequence.right)};
      },

      [&](const rules::Repeat &repeat) {
        return rules::Repeat{apply(*repeat.rule)};
      },

      [&](const rules::Metadata &metadata) {
        return rules::Metadata{apply(*metadata.rule), metadata.params};
      },

      [](auto) {
        assert(!"Unexpected rule type");
        return rules::Blank{};
      }
    );
  }

  Symbol intern_symbol(rules::NamedSymbol named_symbol) {
    for (size_t i = 0; i < grammar.variables.size(); i++) {
      if (grammar.variables[i].name == named_symbol.value) {
        return Symbol::non_terminal(i);
      }
    }

    for (size_t i = 0; i < grammar.external_tokens.size(); i++) {
      if (grammar.external_tokens[i] == named_symbol) {
        return Symbol::external(i);
      }
    }

    missing_rule_name = named_symbol.value;
    return rules::NONE();
  }

  explicit SymbolInterner(const InputGrammar &grammar) : grammar(grammar) {}
  const InputGrammar &grammar;
  string missing_rule_name;
};

CompileError missing_rule_error(string rule_name) {
  return CompileError(TSCompileErrorTypeUndefinedSymbol,
                      "Undefined rule '" + rule_name + "'");
}

pair<InternedGrammar, CompileError> intern_symbols(const InputGrammar &grammar) {
  InternedGrammar result;

  SymbolInterner interner(grammar);

  for (const Rule &external_token : grammar.external_tokens) {
    string external_token_name;
    VariableType external_token_type = VariableTypeAnonymous;
    external_token.match(
      [&](rules::NamedSymbol named_symbol) {
        external_token_name = named_symbol.value;
        if (external_token_name[0] == '_') {
          external_token_type = VariableTypeHidden;
        } else {
          external_token_type =VariableTypeNamed;
        }
      },
      [](auto rule) {}
    );

    auto new_rule = interner.apply(external_token);
    if (!interner.missing_rule_name.empty()) {
      return { result, missing_rule_error(interner.missing_rule_name) };
    }

    result.external_tokens.push_back(Variable{
      external_token_name,
      external_token_type,
      new_rule,
    });
  }

  if (grammar.variables[0].name[0] == '_') {
    return {result, CompileError(TSCompileErrorTypeInvalidRuleName, "A grammar's start rule must be visible.")};
  }

  for (auto &variable : grammar.variables) {
    auto new_rule = interner.apply(variable.rule);
    if (!interner.missing_rule_name.empty()) {
      return { result, missing_rule_error(interner.missing_rule_name) };
    }

    result.variables.push_back(Variable{
      variable.name,
      variable.name[0] == '_' ? VariableTypeHidden : VariableTypeNamed,
      new_rule
    });
  }

  for (auto &rule : grammar.extra_tokens) {
    auto new_rule = interner.apply(rule);
    if (!interner.missing_rule_name.empty()) {
      return { result, missing_rule_error(interner.missing_rule_name) };
    }
    result.extra_tokens.push_back(new_rule);
  }

  for (auto &expected_conflict : grammar.expected_conflicts) {
    set<rules::Symbol> entry;
    for (auto &named_symbol : expected_conflict) {
      auto symbol = interner.intern_symbol(named_symbol);
      if (symbol != rules::NONE()) {
        entry.insert(symbol);
      }
    }
    result.expected_conflicts.insert(entry);
  }

  for (auto &named_symbol : grammar.variables_to_inline) {
    auto symbol = interner.intern_symbol(named_symbol);
    if (symbol != rules::NONE()) {
      result.variables_to_inline.insert(symbol);
    }
  }

  result.word_token = interner.intern_symbol(grammar.word_token);

  return {result, CompileError::none()};
}

}  // namespace prepare_grammar
}  // namespace tree_sitter

File: src/prepare_grammar/intern_symbols.rs
use crate::error::{Error, Result};
use crate::rules::{Rule, Symbol};
use crate::grammars::{InputGrammar, InputVariable, VariableType};
use std::rc::Rc;
use super::InternedGrammar;

pub(super) fn intern_symbols(grammar: &InputGrammar) -> Result<InternedGrammar> {
    let interner = Interner { grammar };

    if variable_type_for_name(&grammar.variables[0].name) == VariableType::Hidden {
        return Err(Error::GrammarError("Grammar's start rule must be visible".to_string()));
    }

    let mut variables = Vec::with_capacity(grammar.variables.len());
    for variable in grammar.variables.iter() {
        variables.push(InputVariable {
            name: variable.name.clone(),
            kind: variable_type_for_name(&variable.name),
            rule: interner.intern_rule(&variable.rule)?,
        });
    }

    let mut external_tokens = Vec::with_capacity(grammar.external_tokens.len());
    for external_token in grammar.external_tokens.iter() {
        let rule = interner.intern_rule(&external_token)?;
        let (name, kind) = if let Rule::NamedSymbol(name) = external_token {
            (name.clone(), variable_type_for_name(&name))
        } else {
            (String::new(), VariableType::Anonymous)
        };
        external_tokens.push(InputVariable { name, kind, rule });
    }

    let mut extra_tokens = Vec::with_capacity(grammar.extra_tokens.len());
    for extra_token in grammar.extra_tokens.iter() {
        extra_tokens.push(interner.intern_rule(extra_token)?);
    }

    let mut expected_conflicts = Vec::new();
    for conflict in grammar.expected_conflicts.iter() {
        let mut interned_conflict = Vec::with_capacity(conflict.len());
        for name in conflict {
            interned_conflict.push(interner
                .intern_name(&name)
                .ok_or_else(|| symbol_error(name))?
            );
        }
        expected_conflicts.push(interned_conflict);
    }

    let mut variables_to_inline = Vec::new();
    for name in grammar.variables_to_inline.iter() {
        if let Some(symbol) = interner.intern_name(&name) {
            variables_to_inline.push(symbol);
        }
    }

    let mut word_token = None;
    if let Some(name) = grammar.word_token.as_ref() {
        word_token = Some(interner
            .intern_name(&name)
            .ok_or_else(|| symbol_error(&name))?
        );
    }

    Ok(InternedGrammar {
        variables,
        external_tokens,
        extra_tokens,
        expected_conflicts,
        variables_to_inline,
        word_token,
    })
}

struct Interner<'a> {
    grammar: &'a InputGrammar
}

impl<'a> Interner<'a> {
    fn intern_rule(&self, rule: &Rule) -> Result<Rule> {
        match rule {
            Rule::Choice { elements } => {
                let mut result = Vec::with_capacity(elements.len());
                for element in elements {
                    result.push(self.intern_rule(element)?);
                }
                Ok(Rule::Choice { elements: result })
            },

            Rule::Seq { left, right } =>
                Ok(Rule::Seq {
                    left: Rc::new(self.intern_rule(left)?),
                    right: Rc::new(self.intern_rule(right)?),
                }),

            Rule::Repeat(content) =>
                Ok(Rule::Repeat(Rc::new(self.intern_rule(content)?))),

            Rule::Metadata { rule, params } =>
                Ok(Rule::Metadata {
                    rule: Rc::new(self.intern_rule(rule)?),
                    params: params.clone()
                }),

            Rule::NamedSymbol(name) => {
                if let Some(symbol) = self.intern_name(&name) {
                    Ok(Rule::Symbol(symbol))
                } else {
                    Err(symbol_error(name))
                }
            },

            _ => Ok(rule.clone())

        }
    }

    fn intern_name(&self, symbol: &str) -> Option<Symbol> {
        for (i, variable) in self.grammar.variables.iter().enumerate() {
            if variable.name == symbol {
                return Some(Symbol::non_terminal(i))
            }
        }

        for (i, external_token) in self.grammar.external_tokens.iter().enumerate() {
            if let Rule::NamedSymbol(name) = external_token {
                if name == symbol {
                    return Some(Symbol::external(i))
                }
            }
        }

        return None
    }
}

fn symbol_error(name: &str) -> Error {
    Error::SymbolError(format!("Undefined symbol '{}'", name))
}

fn variable_type_for_name(name: &str) -> VariableType {
    if name.starts_with("_") {
        VariableType::Hidden
    } else {
        VariableType::Named
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_repeat_expansion() {
        let grammar = intern_symbols(&build_grammar(vec![
            InputVariable::named("x", Rule::choice(vec![
                Rule::named("y"),
                Rule::named("_z"),
            ])),
            InputVariable::named("y", Rule::named("_z")),
            InputVariable::named("_z", Rule::string("a")),
        ])).unwrap();

        assert_eq!(grammar.variables, vec![
            InputVariable::named("x", Rule::choice(vec![
                Rule::non_terminal(1),
                Rule::non_terminal(2),
            ])),
            InputVariable::named("y", Rule::non_terminal(2)),
            InputVariable::hidden("_z", Rule::string("a")),
        ]);
    }

    #[test]
    fn test_interning_external_token_names() {
        // Variable `y` is both an internal and an external token.
        // Variable `z` is just an external token.
        let mut input_grammar = build_grammar(vec![
            InputVariable::named("w", Rule::choice(vec![
                Rule::named("x"),
                Rule::named("y"),
                Rule::named("z"),
            ])),
            InputVariable::named("x", Rule::string("a")),
            InputVariable::named("y", Rule::string("b")),
        ]);
        input_grammar.external_tokens.extend(vec![
            Rule::named("y"),
            Rule::named("z"),
        ]);

        let grammar = intern_symbols(&input_grammar).unwrap();

        // Variable `y` is referred to by its internal index.
        // Variable `z` is referred to by its external index.
        assert_eq!(grammar.variables, vec![
            InputVariable::named("w", Rule::choice(vec![
                Rule::non_terminal(1),
                Rule::non_terminal(2),
                Rule::external(1),
            ])),
            InputVariable::named("x", Rule::string("a")),
            InputVariable::named("y", Rule::string("b")),
        ]);

        // The external token for `y` refers back to its internal index.
        assert_eq!(grammar.external_tokens, vec![
            InputVariable::named("y", Rule::non_terminal(2)),
            InputVariable::named("z", Rule::external(1)),
        ]);
    }

    #[test]
    fn test_grammar_with_undefined_symbols() {
        let result = intern_symbols(&build_grammar(vec![
            InputVariable::named("x", Rule::named("y")),
        ]));

        match result {
            Err(Error::SymbolError(message)) => assert_eq!(message, "Undefined symbol 'y'"),
            _ => panic!("Expected an error but got none"),
        }
    }

    fn build_grammar(variables: Vec<InputVariable>) -> InputGrammar {
        InputGrammar {
            variables,
            name: "the_language".to_string(),
            extra_tokens: Vec::new(),
            external_tokens: Vec::new(),
            expected_conflicts: Vec::new(),
            variables_to_inline: Vec::new(),
            word_token: None,
        }
    }
}
